syntax = "proto3";

package Serialize;

message TradeOrder {
    enum TradeType {
        BUY = 0;
        SELL = 1;
    }
    TradeType type = 1;
    double usd_cost = 2;
    double usd_amount = 3;
}

message TradeRequest {
    enum CommandType {
        MAKE_ORDER = 0;
        VIEW_ACTIVE_ORDERS = 1;
        VIEW_CONPLETED_TRADES = 2;
        VIEW_QUOTE_HISTORY = 3;
        CANCEL_ACTIVE_ORDERS = 4;
    }
    CommandType command = 1;
    TradeOrder order = 2;
    string username = 3;
}

//////////////////////////////////////////////////

message TradeResponse {
    enum status {
        ORDER_SUCCESSFULLY_CREATED = 0;
        ERROR = 1;
    }
    status response_msg = 1;
    AccountBalance account_balance = 2;
    repeated TradeOrder active_orders = 3;
    repeated TradeOrder completed_orders = 4;
    QuoteHistory quote_history = 5;
}

message AccountBalance {
    double usd_balance = 1;
    double rub_balance = 2;
}

message QuoteHistory {
    repeated Quote quotes = 1;
}

message Quote {
    int32 timestamp = 1;
    double price = 2;
}
